### PR TITLE
Fixes #14572 - Constrains JobView (and related views) badge to specific named job

### LINKED ISSUE #14572: Script job count is incorrect
### Deployment Type

Self-hosted

### NetBox Version

v3.6.7

### Python Version

3.10

### Steps to Reproduce

1. Create a script file:

```python
from extras.scripts import Script


class TestScript(Script):
    name = 'Test'
    description = ''

    def run(self, data, commit=True):
        self.log_success('Finished Script')


class TestScriptNew(Script):
    name = 'Test New'
    description = ''

    def run(self, data, commit=True):
        self.log_success('Finished Script 2')
```
2. Upload to NetBox
3. Run `Test`
5. View `Test New`

### Expected Behavior

Job tab badge has "0"

### Observed Behavior

Job tab badge has >0

### PR DESCRIPTION
**Pull request by DanSheps**
### Fixes: #14572 - Constrains JobView to specific job

* Create new method to return relevant jobs
* Call method to provide 'jobs' key for rendered context data

### CODE REVIEW

File: netbox/extras/views.py [hunk 1]
```diff
@@ -1007,6 +1007,23 @@ def post(self, request, id):
         return redirect(reverse('home'))
 
 
+#
+# Function for returning job data - used in scripts and reports
+#
+
+# Return jobs filtered for job model type (Example: `scriptmodule`)
+def get_jobs(module, job_object, jobtype):
```
Comment (jeremystretch) [inline netbox/extras/views.py:1015]: Can we move this to a method on the Report & Script classes? That would keep the code closer to the models, and obviate needing to pass `jobs` as context under every view.

(If not, the function should at least be moved to `utils.py`.)

File: netbox/templates/extras/script/base.html [hunk 2]
```diff
@@ -33,7 +33,7 @@
     </li>
     <li class="nav-item" role="presentation">
       <a class="nav-link{% if tab == 'jobs' %} active{% endif %}" href="{% url 'extras:script_jobs' module=script.module name=script.class_name %}">
-        {% trans "Jobs" %} {% badge module.jobs.count %}
+        {% trans "Jobs" %} {%  badge jobs.count %}
```
Comment (jeremystretch) [inline netbox/templates/extras/script/base.html:36]:
```suggestion
        {% trans "Jobs" %} {% badge jobs.count %}
```

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/extras/views.py (hunk 1 above)
Comment (DanSheps) [reply netbox/extras/views.py:1015]: Yup, definitely, I figured we wanted to reduce the reproduction.  Are you good if I make a Mixin for the classes?

File: netbox/extras/views.py [hunk 3]
```diff
@@ -1057,16 +1057,14 @@ def get_required_permission(self):
...
         module = get_report_module(module, request)
         report = module.reports[name]()
+        jobs = module.get_jobs(name)
 
-        object_type = ContentType.objects.get(app_label='extras', model='reportmodule')
-        report.result = Job.objects.filter(
-            object_type=object_type,
-            object_id=module.pk,
-            name=report.name,
+        report.result = jobs.exclude(
```
Comment (jeremystretch) [inline netbox/extras/views.py:1062]: Shouldn't this be `filter()`?
```suggestion
        report.result = jobs.filter(
```

File: netbox/extras/views.py [hunk 4]
```diff
@@ -1057,16 +1057,14 @@ def get_required_permission(self):
     def get(self, request, module, name):
         module = get_report_module(module, request)
         report = module.reports[name]()
+        jobs = module.get_jobs(name)
```
Comment (jeremystretch) [inline netbox/extras/views.py:1060]: I think this should pass `report.class_name` as below.

File: netbox/extras/views.py [hunk 5]
```diff
@@ -1133,13 +1136,9 @@ def get_required_permission(self):
...
 
         object_type = ContentType.objects.get(app_label='extras', model='reportmodule')
```
Comment (jeremystretch) [inline netbox/extras/views.py:1141]: Doesn't appear to be needed anymore.
```suggestion
```

File: netbox/extras/views.py [hunk 6]
```diff
@@ -1133,13 +1136,9 @@ def get_required_permission(self):
     def get(self, request, module, name):
         module = get_report_module(module, request)
         report = module.reports[name]()
+        jobs = module.get_jobs(name)
```
Comment (jeremystretch) [inline netbox/extras/views.py:1139]: The prior code is passing `report.class_name`, which IIRC is correct.

File: netbox/extras/views.py [hunk 7]
```diff
@@ -1309,13 +1310,7 @@ def get_required_permission(self):
...
         module = get_script_module(module, request)
         script = module.scripts[name]()
-
-        object_type = ContentType.objects.get(app_label='extras', model='scriptmodule')
-        jobs = Job.objects.filter(
-            object_type=object_type,
-            object_id=module.pk,
-            name=script.class_name
-        )
+        jobs = module.get_jobs(name)
```
Comment (jeremystretch) [inline netbox/extras/views.py:1313]:
```suggestion
        jobs = module.get_jobs(script.class_name)
```

File: netbox/extras/views.py [hunk 8]
```diff
@@ -1232,19 +1232,16 @@ def get_required_permission(self):
     def get(self, request, module, name):
         module = get_script_module(module, request)
         script = module.scripts[name]()
+        jobs = module.get_jobs(name)
```
Comment (jeremystretch) [inline netbox/extras/views.py:1235]:
```suggestion
        jobs = module.get_jobs(script.class_name)
```

File: netbox/extras/models/mixins.py [hunk 9]
```diff
@@ -8,6 +8,11 @@
 
 class PythonModuleMixin:
 
+    def get_jobs(self, name):
```
Comment (jeremystretch) [inline netbox/extras/models/mixins.py:11]: Could you add a short docstring here indicating the method's purpose?

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Looks good, but we need to be careful about passing the report/script _class_ name rather than its proper name when fetching jobs.

**Review by jeremystretch — APPROVED** [review]
Thanks @DanSheps!

**Pull request merged**